performance-metrics/src/stats.rs

```diff
@@ -0,0 +1,311 @@
+use log::info;
+use std::cell::RefCell;
+use std::cmp::max;
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::RwLock;
+use std::task::Poll;
+use std::time::{Duration, Instant};
+
+use actix;
+use futures;
+use futures::task::Context;
+use std::mem;
+use std::pin::Pin;
+
+pub static NTHREADS: AtomicUsize = AtomicUsize::new(0);
+const SLOW_CALL_THRESHOLD: Duration = Duration::from_millis(500);
+const MIN_OCCUPANCY_RATIO_THRESHOLD: f64 = 0.00;
+#[cfg(feature = "performance_stats")]
+const MESSAGE_LIMIT: usize = 250;
+
+lazy_static! {
+    static ref STATS: RwLock<Stats> = RwLock::new(Stats::new());
+    static ref REF_COUNTER: RwLock<HashMap<(&'static str, u32), u128>> =
+        RwLock::new(HashMap::new());
+}
+
+thread_local! {
+    static TID: RefCell<usize> = RefCell::new(0);
+}
+
+#[derive(Default)]
+struct Entry {
+    cnt: u128,
+    time: Duration,
+    max_time: Duration,
+}
+
+struct ThreadStats {
+    stat: HashMap<(&'static str, u32), Entry>,
+    cnt: u128,
+    time: Duration,
+    classes: HashSet<&'static str>,
+}
+
+impl ThreadStats {
+    fn new() -> Self {
+        Self { stat: HashMap::new(), cnt: 0, time: Duration::default(), classes: HashSet::new() }
+    }
+
+    fn log(&mut self, class_name: &'static str, msg: &'static str, line: u32, took: Duration) {
+        let entry = self.stat.entry((msg, line)).or_insert_with(|| Entry {
+            cnt: 0,
+            time: Duration::default(),
+            max_time: Duration::default(),
+        });
+        entry.cnt += 1;
+        entry.time += took;
+        entry.max_time = max(took, entry.max_time);
+
+        self.cnt += 1;
+        self.time += took;
+        self.classes.insert(class_name);
+    }
+
+    fn print_stats(&self, tid: usize, sleep_time: Duration) {
+        let ratio = (self.time.as_nanos() as f64) / (sleep_time.as_nanos() as f64);
+
+        let class_name = format!("{:?}", self.classes);
+
+        if ratio > MIN_OCCUPANCY_RATIO_THRESHOLD {
+            info!(
```

Comment:
As far as I remember, we need to specify `target: "performance_metric"` in every log call, so tracing filters apply correctly.

Reply:
Yep

Reply:
Yes, I did it in config.rs